**Load the Dataset**

In [1]:
import pandas as pd

In [6]:
df=pd.read_csv("netflix_titles.csv")

In [7]:
df.head()

show_id     type                  title         director  \
0      s1    Movie   Dick Johnson Is Dead  Kirsten Johnson   
1      s2  TV Show          Blood & Water              NaN   
2      s3  TV Show              Ganglands  Julien Leclercq   
3      s4  TV Show  Jailbirds New Orleans              NaN   
4      s5  TV Show           Kota Factory              NaN   

                                                cast        country  \
0                                                NaN  United States   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   South Africa   
2  Sami Bouajila, Tracy Gotoas, Samuel Jouy, Nabi...            NaN   
3                                                NaN            NaN   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...          India   

           date_added  release_year rating   duration  \
0  September 25, 2021          2020  PG-13     90 min   
1  September 24, 2021          2021  TV-MA  2 Seasons   
2  September 24, 2021          2021  TV-MA   1 Season   
3  September 24, 2021          2021  TV-MA   1 Season   
4  September 24, 2021          2021  TV-MA  2 Seasons   

                                           listed_in  \
0                                      Documentaries   
1    International TV Shows, TV Dramas, TV Mysteries   
2  Crime TV Shows, International TV Shows, TV Act...   
3                             Docuseries, Reality TV   
4  International TV Shows, Romantic TV Shows, TV ...   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
2  To protect his family from a powerful drug lor...  
3  Feuds, flirtations and toilet talk go down amo...  
4  In a city of coaching centers known to train I...

In [8]:
#basic info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8807 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   show_id       8807 non-null   object
 1   type          8807 non-null   object
 2   title         8807 non-null   object
 3   director      6173 non-null   object
 4   cast          7982 non-null   object
 5   country       7976 non-null   object
 6   date_added    8797 non-null   object
 7   release_year  8807 non-null   int64 
 8   rating        8803 non-null   object
 9   duration      8804 non-null   object
 10  listed_in     8807 non-null   object
 11  description   8807 non-null   object
dtypes: int64(1), object(11)
memory usage: 825.8+ KB


In [10]:
#Statistical summary
df.describe(include="all")

show_id   type   title       director                cast  \
count     8807   8807    8807           6173                7982   
unique    8807      2    8807           4528                7692   
top      s8807  Movie  Zubaan  Rajiv Chilaka  David Attenborough   
freq         1   6131       1             19                  19   
mean       NaN    NaN     NaN            NaN                 NaN   
std        NaN    NaN     NaN            NaN                 NaN   
min        NaN    NaN     NaN            NaN                 NaN   
25%        NaN    NaN     NaN            NaN                 NaN   
50%        NaN    NaN     NaN            NaN                 NaN   
75%        NaN    NaN     NaN            NaN                 NaN   
max        NaN    NaN     NaN            NaN                 NaN   

              country       date_added  release_year rating  duration  \
count            7976             8797   8807.000000   8803      8804   
unique            748             1767           NaN     17       220   
top     United States  January 1, 2020           NaN  TV-MA  1 Season   
freq             2818              109           NaN   3207      1793   
mean              NaN              NaN   2014.180198    NaN       NaN   
std               NaN              NaN      8.819312    NaN       NaN   
min               NaN              NaN   1925.000000    NaN       NaN   
25%               NaN              NaN   2013.000000    NaN       NaN   
50%               NaN              NaN   2017.000000    NaN       NaN   
75%               NaN              NaN   2019.000000    NaN       NaN   
max               NaN              NaN   2021.000000    NaN       NaN   

                           listed_in  \
count                           8807   
unique                           514   
top     Dramas, International Movies   
freq                             362   
mean                             NaN   
std                              NaN   
min                              NaN   
25%                              NaN   
50%                              NaN   
75%                              NaN   
max                              NaN   

                                              description  
count                                                8807  
unique                                               8775  
top     Paranormal activity at a lush, abandoned prope...  
freq                                                    4  
mean                                                  NaN  
std                                                   NaN  
min                                                   NaN  
25%                                                   NaN  
50%                                                   NaN  
75%                                                   NaN  
max                                                   NaN

**Idetityfy missing values**

In [11]:
df.isnull().sum()

show_id            0
type               0
title              0
director        2634
cast             825
country          831
date_added        10
release_year       0
rating             4
duration           3
listed_in          0
description        0
dtype: int64

- Based on this function I found the null values such as director,cast,country,date_added,rating,duration

In [12]:
#fill missing 'director' and 'cast' with 'Not Available'
df['director'] = df['director'].fillna('Not Available')
df['cast'] = df['cast'].fillna('Not Available')

# Drop rows where 'country' or 'date_added' is missing
df = df.dropna(subset=['country', 'date_added'])

# Fill missing 'rating' with most frequent value
df['rating'] = df['rating'].fillna(df['rating'].mode()[0])
# Drop or fill missing 'duration'
df['duration'] = df['duration'].fillna('Unknown')

In [13]:
# Check if all missing values are handled
print(df.isnull().sum())

show_id         0
type            0
title           0
director        0
cast            0
country         0
date_added      0
release_year    0
rating          0
duration        0
listed_in       0
description     0
dtype: int64


In [15]:
#count duplicate vaues
df.duplicated().sum()

np.int64(0)

**Standardise text values**

In [16]:
# Strip and lowercase certain fields
df['type'] = df['type'].str.strip().str.lower()
df['country'] = df['country'].str.strip()
df['rating'] = df['rating'].str.strip()

**Fix date format**

In [18]:
df['date_added'] = pd.to_datetime(df['date_added'], errors='coerce')

**Rename coulmns to be consistence**

In [20]:
df.columns = df.columns.str.strip().str.lower().str.replace(' ', '_')

In [21]:
df.head()

show_id     type                          title         director  \
0      s1    movie           Dick Johnson Is Dead  Kirsten Johnson   
1      s2  tv show                  Blood & Water    Not Available   
4      s5  tv show                   Kota Factory    Not Available   
7      s8    movie                        Sankofa     Haile Gerima   
8      s9  tv show  The Great British Baking Show  Andy Devonshire   

                                                cast  \
0                                      Not Available   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   

                                             country date_added  release_year  \
0                                      United States 2021-09-25          2020   
1                                       South Africa 2021-09-24          2021   
4                                              India 2021-09-24          2021   
7  United States, Ghana, Burkina Faso, United Kin... 2021-09-24          1993   
8                                     United Kingdom 2021-09-24          2021   

  rating   duration                                          listed_in  \
0  PG-13     90 min                                      Documentaries   
1  TV-MA  2 Seasons    International TV Shows, TV Dramas, TV Mysteries   
4  TV-MA  2 Seasons  International TV Shows, Romantic TV Shows, TV ...   
7  TV-MA    125 min   Dramas, Independent Movies, International Movies   
8  TV-14  9 Seasons                       British TV Shows, Reality TV   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
4  In a city of coaching centers known to train I...  
7  On a photo shoot in Ghana, an American model s...  
8  A talented batch of amateur bakers face off in...

**Fianal Check**

In [23]:
df.info()
df.head()

<class 'pandas.core.frame.DataFrame'>
Index: 7967 entries, 0 to 8806
Data columns (total 12 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   show_id       7967 non-null   object        
 1   type          7967 non-null   object        
 2   title         7967 non-null   object        
 3   director      7967 non-null   object        
 4   cast          7967 non-null   object        
 5   country       7967 non-null   object        
 6   date_added    7882 non-null   datetime64[ns]
 7   release_year  7967 non-null   int64         
 8   rating        7967 non-null   object        
 9   duration      7967 non-null   object        
 10  listed_in     7967 non-null   object        
 11  description   7967 non-null   object        
dtypes: datetime64[ns](1), int64(1), object(10)
memory usage: 809.1+ KB


show_id     type                          title         director  \
0      s1    movie           Dick Johnson Is Dead  Kirsten Johnson   
1      s2  tv show                  Blood & Water    Not Available   
4      s5  tv show                   Kota Factory    Not Available   
7      s8    movie                        Sankofa     Haile Gerima   
8      s9  tv show  The Great British Baking Show  Andy Devonshire   

                                                cast  \
0                                      Not Available   
1  Ama Qamata, Khosi Ngema, Gail Mabalane, Thaban...   
4  Mayur More, Jitendra Kumar, Ranjan Raj, Alam K...   
7  Kofi Ghanaba, Oyafunmike Ogunlano, Alexandra D...   
8  Mel Giedroyc, Sue Perkins, Mary Berry, Paul Ho...   

                                             country date_added  release_year  \
0                                      United States 2021-09-25          2020   
1                                       South Africa 2021-09-24          2021   
4                                              India 2021-09-24          2021   
7  United States, Ghana, Burkina Faso, United Kin... 2021-09-24          1993   
8                                     United Kingdom 2021-09-24          2021   

  rating   duration                                          listed_in  \
0  PG-13     90 min                                      Documentaries   
1  TV-MA  2 Seasons    International TV Shows, TV Dramas, TV Mysteries   
4  TV-MA  2 Seasons  International TV Shows, Romantic TV Shows, TV ...   
7  TV-MA    125 min   Dramas, Independent Movies, International Movies   
8  TV-14  9 Seasons                       British TV Shows, Reality TV   

                                         description  
0  As her father nears the end of his life, filmm...  
1  After crossing paths at a party, a Cape Town t...  
4  In a city of coaching centers known to train I...  
7  On a photo shoot in Ghana, an American model s...  
8  A talented batch of amateur bakers face off in...

**Save cleaned dataset**

In [24]:
df.to_csv("netflix_cleaned.csv", index=False)

**Data Cleaning Summary – Netflix Titles Dataset**
- Dataset: netflix_titles.csv
- Total Rows Before Cleaning: 8807

**Cleaning Steps Performed**
- Handled Missing Values
- Replaced 2634 missing entries in the director column with "Not Available".
- Replaced 825 missing entries in the cast column with "Not Available".
- Dropped 831 rows with missing country values, as location is crucial for analysis.
- Dropped 10 rows with missing date_added, required for time-based insights.
- Filled 4 missing rating values with the most frequent rating (mode).
- Filled 3 missing duration values with "Unknown".

**Removed Duplicates**
- Dropped all exact duplicate rows using .drop_duplicates().
- Standardized Text Data
- Stripped whitespaces and converted text columns like type, country, and rating to lowercase for consistency.
  
**Fixed Date Format**
- Converted the date_added column to datetime format using pd.to_datetime().
  
**Renamed Columns**
- Renamed all column headers to lowercase with underscores instead of spaces for uniformity and easier access in code.

**Final Output**
- Cleaned Rows Remaining: ~7966
- File Saved As: netflix_cleaned.csv



**Interview questions and answers**

**1. What are missing values and how do you handle them?**
- Missing values occur when no data is stored for a variable in an observation.
#### Handling techniques:
- Remove rows/columns with .dropna()
- Impute missing data using .fillna():
- Mean/median for numerical data
- Mode or placeholder (e.g., "Unknown") for categorical data
- Domain-specific methods, like predictive imputation

**2. How do you treat duplicate records?**
- Duplicate records are identical rows that appear more than once.
#### Handling method in Pandas:
- df.drop_duplicates(inplace=True)

**3. Difference between dropna() and fillna() in Pandas?**
- dropna() removes rows or columns with missing values.
- df.dropna(subset=['column_name'])
- fillna() fills missing values with a specified value.
- df['column_name'].fillna('Unknown', inplace=True)
    
**4. What is outlier treatment and why is it important?**
- Outliers are data points significantly different from others.
#### Treatment methods:
- Remove outliers using IQR or Z-score method
- Cap/floor values (e.g., winsorizing)
- Log transformation or scaling
#### Importance:
- Outliers can distort analysis, affect mean/variance, and reduce model performance.

**5. Explain the process of standardizing data.**
- Standardization ensures consistency across the dataset.
#### It includes:
- Converting text to lowercase
- Removing leading/trailing spaces
- Unifying categorical labels (e.g., "Male" vs "male")

#### Date formatting
- Renaming columns (snake_case, no spaces)

**6. How do you handle inconsistent data formats (e.g., date/time)?**
- Use Pandas to_datetime() for date formats:
- df['date_column'] = pd.to_datetime(df['date_column'], errors='coerce')
- This converts all dates to a standard format and marks invalid ones as NaT.

**7. What are common data cleaning challenges?**
- Inconsistent formats (e.g., "USA" vs "United States")
- Missing or invalid data
- Duplicate entries
- Hidden nulls (like "N/A", "?", "unknown")
- Incorrect data types
    
**8. How can you check data quality?**
- df.info() → Check data types and nulls
- df.isnull().sum() → Count missing values
- df.duplicated().sum() → Check for duplicate rows
- df.value_counts() → Spot inconsistent categorical values
- df.describe() → Review summary stats for anomalies
